### PR TITLE
Update ui_extension to support Unified config

### PR DESCRIPTION
**Pull request by isaacroldan**
### WHY are these changes introduced?
Adds support for the new UI Extensions schema with the `targeting` field.

### WHAT is this pull request doing?
- Update `ui_extension` schema to support both `extension_instance` and `targeting`.
- Internally keep working with `extension_instance`
- Update `ui_extension` template to follow unified config

### How to test your changes?
- From main generate a new `ui_extension` (it'll have the old format)
- From this branch, and on the same app, generate another `ui_extension (should have the new format)
- `app info` should load both extensions without problems.

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/models/extensions/specifications/ui_extension.ts [hunk 1]
```diff
@@ -20,13 +27,65 @@ const UIExtensionSchema = BaseSchema.extend({
...
+  description: zod.string().optional(),
+  type: zod.literal('ui_extension'),
+  extensions: zod.array(UIExtensionSchema).min(1).max(1),
+  settings: UnifiedSettingsSchema,
+}).transform((config) => {
+  const newConfig = {
+    ...config,
+    name: config.name,
+    type: config.type,
+    extension_points: getArrayRejectingUndefined(config.extensions[0]?.targeting ?? []),
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/specifications/ui_extension.ts:71]: the apiVersion is part of the `BaseSchema` so it should be part of `...config` already, or should it be somewhere else?

[the remark this answers is not in the document]
Comment (vividviolet) [reply packages/app/src/cli/models/extensions/specifications/ui_extension.ts:70]: Does it inherit from the global? `apiVersion`, `name` and `description` should all inherit from the global config if it's not explicitly set

**Review by vividviolet — COMMENTED** [review]
@lemonmade @davejcameron Should we support inheriting metafields and settings from the global?

**Comment by vividviolet** [comment]
Ok, I thought Function had `metafields` too but it turned out to be `input.metafields`: https://shopify.dev/docs/apps/functions/input-output/variables-queries#step-1-specify-the-metafield-to-use

@isaacroldan I think in summary these are all the fields that should be present, some of which need to be inherited either from the top-level or the extension-level.

```toml
api_version = "2023-04"
name = "Google maps validation"
description = "Validates location on Google maps"

[[extensions]]
handle = "google-maps-function"
type = "function"
# inherits top-level global api_version if not set
# inherits top-level global name if not set
# inherits top-level global description if not set
  [extensions.build]
  command = "cargo wasi build --release"
  path = "google-maps-function/dist/function.wasm"

  [[extensions.targeting]]
  export = "fetch"
  input_query = "google-maps-function/fetch.input.graphql"
  target = "purchase.validation.fetch"

  [[extensions.targeting]]
  export = "validate"                                         
  input_query = "google-maps-function/fetch.validate.graphql"  target = "purchase.validation.validate"

[[extensions]]
handle = "google-maps-ui"
type = "ui-extension"
# inherits top-level global api_version if not set
# inherits top-level global name if not set
# inherits top-level global description if not set

[[metafields]]
namespace = "my-namespace"
key = "my-key"

  [[extensions.targeting]]
  module = "google-maps-ui/settings.jsx"
  target = "purchase.validation.render-settings"
  # inherits extension-level metafields if not set
```

File: packages/app/src/cli/models/extensions/specifications/ui_extension.ts [hunk 2]
```diff
@@ -20,13 +27,73 @@ const UIExtensionSchema = BaseSchema.extend({
...
+}).transform((config) => {
+  const extensionPoints = config.extensions[0]?.targeting.map((targeting) => {
+    return {
+      target: targeting.target,
+      module: targeting.module,
+      metafields: targeting.metafields ?? config.extensions[0]?.metafields ?? [],
+    }
+  })
+  const newConfig: UIExtensionLegacySchemaType = {
+    name: config.extensions[0]?.name ?? config.name,
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/specifications/ui_extension.ts:76]: We are assuming here that this is for a single ui_extension, a workaround just for ui_extension while we implement the real support for unified config in the whole CLI. 
But David has a PR with a prototype of that, so we might discard parts of this PR

**Comment by isaacroldan** [comment]
Updated! The `ui_extension` schema now supports either a `targeting` or a `extension_points` array.

If using `targeting` then it is transformed to `extension_points` internally. Extension level metafields are inherited in the target if not present.

All other top-level fields are flattened and included in the config of each extension in the array when parsing it. (name, description, api_version...)

I'll update the tests later today

File: packages/app/templates/ui-extensions/projects/ui_extension/shopify.ui.extension.toml.liquid [hunk 3]
```diff
@@ -1,17 +1,20 @@
...
 api_version = "2022-10"
 
 [capabilities]
 api_access = true
 network_access = true
 block_progress = true
 
-[[extension_points]]
-target = "Checkout::Dynamic::Render"
+[[extensions]]
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/templates/ui-extensions/projects/ui_extension/shopify.ui.extension.toml.liquid:9]: Is it safe If I remove it in this PR then?

File: packages/app/src/cli/models/extensions/specifications/ui_extension.test.ts [hunk 4]
```diff
@@ -65,6 +66,70 @@ describe('ui_extension', async () => {
...
+        targeting: [
+          {
+            target: 'EXTENSION::POINT::A',
+            module: './src/ExtensionPointA.js',
+          },
+        ],
+        api_version: '2023-01' as const,
+        name: 'UI Extension',
+        type: 'ui_extension',
+        metafields: [{namespace: 'test', key: 'test'}],
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/specifications/ui_extension.test.ts:82]: It's not included in the deployConfig for sure, not sure about how we build the `dev` payload, but I didn't change that 🤔

**Comment by vividviolet** [comment]
@isaacroldan you're correct, I got the TOML wrong. Using this works as expected
```toml
name = "action-extension"
api_version = "internal"
[[extensions]]
  type = "ui_extension"
  [[extensions.metafields]]
  namespace = "my-namespace"
  key = "my-key"
  [[extensions.metafields]]
  namespace = "my-namespace"
  key = "my-other-key"
  [capabilities]
  network_access = true
  block_progress = true
  api_access = true
  [settings]
    [[settings.fields]]
    key = "field_key"
    type = "boolean"
    name = "field-name"
    [[settings.fields]]
    key = "field_key_2"
    type = "number_integer"
    name = "field-name-2"
  [[extensions.targeting]]
  target = "admin.product.item.block"
  module = "./index.jsx"
  [[extensions.targeting]]
  target = "admin.product.index.action"
  module = "./index.jsx"
  [[extensions.targeting.metafields]]
  namespace = "my-specific-namespace"
  key = "my-specific-key"
```

I think the only thing left to do is to drop support for setting the metafields at the top-level as this is a ui-extension concept only

File: packages/app/templates/ui-extensions/projects/ui_extension/shopify.ui.extension.toml.liquid (hunk 3 above)
[the remark this answers is not in the document]
Comment (vividviolet) [reply packages/app/templates/ui-extensions/projects/ui_extension/shopify.ui.extension.toml.liquid:9]: You'd put it [here](https://github.com/Shopify/ui-extensions-templates) and then update Partner's to make it available: https://github.com/Shopify/partners/blob/master/app/operations/apps/templates/specifications/v304600.rb

**Review by jamesvidler — COMMENTED** [review]
I tophatted this by generating a new extension with an app I already had in spin (and an existing extension) and it generated the new extension with the new config just fine. 

I was also able to run `pnpm shopify app dev` and see the extension render. 👍 

Screenshot 2023-07-13 at 4 23 20 PM

What is strange though is that the other extension I have appears to be excluded now and ignored. Any idea what might be going on there? It's entirely possible the app structure is out of date, as I did not generate a new app. 

Screenshot 2023-07-13 at 4 30 55 PM

This is the[ branch of the app](https://github.com/Shopify/checkout-ui-extension-dev/tree/jkv/config-updates-test/extensions) I was using to test.

**Review by jamesvidler — APPROVED** [review]

**Pull request merged**